Avoid opening copied wallet databases simultaneously

ryanofsky opened the pull request:
Make sure wallet databases have unique fileids. If they don't, throw an error. BDB caches do not work properly when more than one open database has the same fileid, because values written to one database may show up in reads to other databases.

Bitcoin will never create different databases with the same fileid, but users can create them by manually copying database files.

BDB caching bug was reported by @dooglus in https://github.com/bitcoin/bitcoin/issues/11429

ryanofsky committed:
Avoid opening copied wallet databases simultaneously

Make sure wallet databases have unique fileids. If they don't, throw an error.
BDB caches do not work properly when more than one open database has the same
fileid, because values written to one database may show up in reads to other
databases.

Bitcoin will never create different databases with the same fileid, but users
can create them by manually copying database files.

BDB caching bug was reported by Chris Moore <dooglus@gmail.com>
https://github.com/bitcoin/bitcoin/issues/11429

Fixes #11429

sipa commented:
Concept ACK

jonasschnelli commented:
Nice clean PR!
utACK ebc7c2ff3c31a3a5377e781d2239dae44e7b602e

MeshCollider commented:
Concept ACK

promag commented on src/wallet/db.cpp:
```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
                             0);
 
             if (ret != 0) {
+                error = strprintf("CDB: Error %d, can't open database %s", ret, strFilename);
```
Use same format as below: `CDB: Can't open database %s (open error %d)"`.

ryanofsky replied:
Didn't do this to avoid changing the existing error message.

promag commented on src/wallet/db.cpp:
```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
+                u_int8_t fileid[DB_FILE_ID_LEN];
+                ret = pdb->get_mpf()->get_fileid(fileid);
+                if (ret != 0) {
+                    error = strprintf("CDB: Can't open database %s (get_fileid error %d)\n", strFilename, ret);
```
Remove newline?

ryanofsky replied:
Good catch, removed in e32c2a51b757ab764ec5dbe8bf04b3fc4aa3cb5f

promag commented on src/wallet/db.cpp:
```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
+                        memcmp(fileid, item_fileid, sizeof(fileid)) == 0) {
+                        const char* item_filename = nullptr;
+                        item.second->get_dbname(&item_filename, nullptr);
+                        error = strprintf("CDB: Can't open database %s (duplicates fileid %s from %s)", strFilename,
```
This error overrides the previous.

promag replied:
Should `+= std::endl +` instead?

ryanofsky replied:
Fixed override in e32c2a51b757ab764ec5dbe8bf04b3fc4aa3cb5f, but I'm not clear on what std::endl comment is referring to.

promag replied:
Nevermind, but the suggestion was to accumulate the errors.

promag commented on test/functional/multiwallet.py:
```diff
@@ -29,6 +30,10 @@ def run_test(self):
         self.assert_start_raises_init_error(0, ['-wallet=w11'], 'Error loading wallet w11. -wallet filename must be a regular file.')
 
+        # should not initialize if one wallet is a copy of another
+        shutil.copyfile(self.options.tmpdir + "/node0/regtest/w2", self.options.tmpdir + "/node0/regtest/w22")
```
Use `os.path.join`?

ryanofsky replied:
Done in e32c2a51b757ab764ec5dbe8bf04b3fc4aa3cb5f

promag commented on src/wallet/db.cpp:
```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
+                // an error. BDB caches do not work properly when more than one
+                // open database has the same fileid (values written to one
+                // database may show up in reads to other databases).
+                //
```
Nit, remove empty line.

ryanofsky replied:
It's intended to be a new paragraph.

promag commented on src/wallet/db.cpp:
```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
+                    error = strprintf("CDB: Can't open database %s (get_fileid error %d)\n", strFilename, ret);
+                }
+
+                for (const auto& item : env->mapDb) {
```
Only `for` when `ret == 0`?

ryanofsky replied:
Done in e32c2a51b757ab764ec5dbe8bf04b3fc4aa3cb5f

promag commented on src/wallet/db.cpp:
```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
+                        error = strprintf("CDB: Can't open database %s (duplicates fileid %s from %s)", strFilename,
+                            HexStr(std::begin(item_fileid), std::end(item_fileid)),
+                            item_filename ? item_filename : "");
+                    }
```
Break on first duplicate?

ryanofsky replied:
Done in e32c2a51b757ab764ec5dbe8bf04b3fc4aa3cb5f

laanwj commented on src/wallet/db.cpp:
```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
                 --env->mapFileUseCount[strFile];
                 strFile = "";
-                throw std::runtime_error(strprintf("CDB: Error %d, can't open database %s", ret, strFilename));
+                throw std::runtime_error(error);
```
I guess the reason for storing the error in a string, and not just raising the exceptions at the place the error occurs, above, is to avoid duplicating this cleanup code?
It seems a bit circuitous, but I understand. An alternative would be to wrap these operations in a RAII operation.

ryanofsky replied:
This was one reason. The other reason was wanting this to be self-contained and not change existing code.

ryanofsky reviewed (commented):
Updated ebc7c2ff3c31a3a5377e781d2239dae44e7b602e -> e32c2a51b757ab764ec5dbe8bf04b3fc4aa3cb5f ([pr/wid.1](https://github.com/ryanofsky/bitcoin/commits/pr/wid.1) -> [pr/wid.2](https://github.com/ryanofsky/bitcoin/commits/pr/wid.2))

promag commented on src/wallet/db.cpp:
```diff
@@ -401,12 +401,48 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
+                // copy database files.
+                for (const auto& item : env->mapDb) {
+                    u_int8_t item_fileid[DB_FILE_ID_LEN];
+                    if (item.second && item.second->get_mpf()->get_fileid(item_fileid) == 0 &&
```
Maybe use the `ret = ...` pattern like above to make easier to read?

ryanofsky commented:
I had to make another change to avoid travis failures. For some reason the get_fileid call was failing with error 22 in accounting tests (strangely only when accounting tests were run in conjunction with other tests, not when running standalone). I fixed it by skipping the get_fileid check on in-memory databases, which actually makes sense anyway.

Updated e32c2a51b757ab764ec5dbe8bf04b3fc4aa3cb5f -> 4164132f6b475d3a3b4f1492f22f515cdb98fcd8 ([pr/wid.2](https://github.com/ryanofsky/bitcoin/commits/pr/wid.2) -> [pr/wid.3](https://github.com/ryanofsky/bitcoin/commits/pr/wid.3))

jnewbery commented:
Tested ACK 4164132f6b475d3a3b4f1492f22f515cdb98fcd8

I think you could improve the code comment by saying that bdb guarantees file identification strings for different databases to be unique, and perhaps linking to the documentation (https://docs.oracle.com/cd/E17275_01/html/programmer_reference/program_copy.html).

ryanofsky commented:
Thanks, updated comment 4164132f6b475d3a3b4f1492f22f515cdb98fcd8 -> f98832222c17c25814b9a18859e72f5041dfb55d ([pr/wid.3](https://github.com/ryanofsky/bitcoin/commits/pr/wid.3) -> [pr/wid.4](https://github.com/ryanofsky/bitcoin/commits/pr/wid.4))

I didn't include the oracle link originally because I wasn't sure if the URL would be stable, but I realize the comment doesn't really make sense without that context, so it seems worth adding.

jnewbery commented:
> I didn't include the oracle link originally because I wasn't sure if the URL would be stable

Yes, I considered that as well. I think it's ok to add the link now. If it dies or changes, then the comment can be updated or removed. For now, it's adding useful additional context.

I can confirm that your new commit updates the comment. It also makes the following change:

```
-    u_int8_t fileid[DB_FILE_ID_LEN] = {0};
+    u_int8_t fileid[DB_FILE_ID_LEN];
```

Was that intentional?

ryanofsky commented:
> Was that intentional?

Yeah, I added the zero initialization when the get_fileid call was failing and I was considering making it optional, but it's not needed now.

jnewbery commented:
Tested ACK f98832222c17c25814b9a18859e72f5041dfb55d

laanwj commented:
Needs rebase for #11492, otherwise seems ready to merge.

ryanofsky commented:
Rebased f98832222c17c25814b9a18859e72f5041dfb55d -> 478a89c1ef79a75275d1b508122c06eee9386b2d ([pr/wid.4](https://github.com/ryanofsky/bitcoin/commits/pr/wid.4) -> [pr/wid.5](https://github.com/ryanofsky/bitcoin/commits/pr/wid.5))
